functions-and-operators/window-functions.md

```diff
@@ -8,22 +8,466 @@ aliases: ['/docs/dev/functions-and-operators/window-functions/','/docs/dev/refer
+
+The `DENSE_RANK()` function is similar to [`RANK()`](#rank) but won't leave any gaps in case of ties.
+
+```sql
+SELECT 
+    *,
+    DENSE_RANK() OVER (ORDER BY n) 
+FROM (
+    SELECT 5 AS 'n'
+    UNION ALL
+    SELECT 8
+    UNION ALL
+    SELECT 5
+    UNION ALL
+    SELECT 30
+    UNION ALL
+    SELECT 31
+    UNION ALL
+    SELECT 32) a;
+```
+
+```
++----+--------------------------------+
+| n  | DENSE_RANK() OVER (ORDER BY n) |
++----+--------------------------------+
+|  5 |                              1 |
+|  5 |                              1 |
+|  8 |                              2 |
+| 30 |                              3 |
+| 31 |                              4 |
+| 32 |                              5 |
++----+--------------------------------+
+6 rows in set (0.00 sec)
+```
+
+## [`FIRST_VALUE()`](https://dev.mysql.com/doc/refman/8.0/en/window-function-descriptions.html#function_first-value)
+
+The `FIRST_VALUE(expr)` returns the first value of a group.
+
+In the following example we use two different window definitions:
+
+1. `PARTITION BY n MOD 2 ORDER BY n` resulting in two groups: `1, 3` and `2, 4`. So this returns either 1 or 2 as those are the first values of those groups.
+2. `PARTITION BY n <= 2 ORDER BY n` resulting in two groups: `1, 2` and `3, 4` So this returns either 1 or 3 depending on which group it is group.
+
+```sql
+SELECT 
+    n,
+    FIRST_VALUE(n) OVER (PARTITION BY n MOD 2 ORDER BY n),
+    FIRST_VALUE(n) OVER (PARTITION BY n <= 2 ORDER BY n)
+FROM (
+    SELECT 1 AS 'n'
+    UNION
+    SELECT 2
+    UNION
+    SELECT 3
+    UNION
+    SELECT 4
+) a
+ORDER BY 
+    n;
+```
+
+```
++------+-------------------------------------------------------+------------------------------------------------------+
+| n    | FIRST_VALUE(n) OVER (PARTITION BY n MOD 2 ORDER BY n) | FIRST_VALUE(n) OVER (PARTITION BY n <= 2 ORDER BY n) |
++------+-------------------------------------------------------+------------------------------------------------------+
+|    1 |                                                     1 |                                                    1 |
+|    2 |                                                     2 |                                                    1 |
+|    3 |                                                     1 |                                                    3 |
+|    4 |                                                     2 |                                                    3 |
++------+-------------------------------------------------------+------------------------------------------------------+
+4 rows in set (0.00 sec)
+```
+
+## [`LAG()`](https://dev.mysql.com/doc/refman/8.0/en/window-function-descriptions.html#function_lag)
+
+The `LAG(expr [, num [, default]])` function returns the value that is lagging behind `num` values behind the the current value. If there is no current value then `default` is returned, which defaults to `NULL`.
```

Comment:
Should "behind `num` values behind the the current value" be "behind `num` values or behind the the current value"?

Reply:
No, I don't think so. The current text looks correct to me.